azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/gremlin/GremlinAutoConfiguration.java

```diff
@@ -0,0 +1,96 @@
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.autoconfigure.domain.EntityScanner;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.data.annotation.Persistent;
+import org.springframework.lang.NonNull;
+import org.springframework.util.ClassUtils;
+
+import java.util.HashMap;
+
+@Configuration
+@ConditionalOnClass({GremlinFactory.class, GremlinTemplate.class, MappingGremlinConverter.class})
+@ConditionalOnProperty(prefix = "gremlin", value = {"endpoint", "port", "username", "password"})
+@EnableConfigurationProperties(GremlinProperties.class)
+public class GremlinAutoConfiguration {
+
+    private final GremlinProperties properties;
+
+    private final TelemetryProxy telemetryProxy;
+
+    private final ApplicationContext applicationContext;
+
+    public GremlinAutoConfiguration(@NonNull GremlinProperties properties, @NonNull ApplicationContext context) {
+        this.properties = properties;
+        this.applicationContext = context;
+        this.telemetryProxy = new TelemetryProxy(properties.isTelemetryAllowed());
+    }
+
+    private void trackCustomEvent() {
+        final HashMap<String, String> customTelemetryProperties = new HashMap<>();
+        final String[] packageNames = this.getClass().getPackage().getName().split("\\.");
+
+        if (packageNames.length > 1) {
+            customTelemetryProperties.put(TelemetryData.SERVICE_NAME, packageNames[packageNames.length - 1]);
+        }
+
+        telemetryProxy.trackEvent(ClassUtils.getUserClass(this.getClass()).getSimpleName(), customTelemetryProperties);
+    }
+
+    @Bean
+    @ConditionalOnMissingBean
+    public GremlinFactory gremlinFactory() {
+        this.trackCustomEvent();
+
+        final String endpoint = this.properties.getEndpoint();
+        final String port = this.properties.getPort();
+        final String username = this.properties.getUsername();
+        final String password = this.properties.getPassword();
+
+        return new GremlinFactory(endpoint, port, username, password);
+    }
+
+    @Bean
+    @ConditionalOnMissingBean
+    public GremlinTemplate gremlinTemplate(GremlinFactory factory, MappingGremlinConverter converter) {
+        return new GremlinTemplate(factory, converter);
+    }
+
+    @Bean
+    @ConditionalOnMissingBean
+    public GremlinMappingContext gremlinMappingContext() {
+        try {
+            final GremlinMappingContext context = new GremlinMappingContext();
+
+            context.setInitialEntitySet(new EntityScanner(this.applicationContext).scan(Persistent.class));
+
+            return context;
+        } catch (ClassNotFoundException e) {
```

Comment:
Which class is possibly not found? Could we use `ConditionalOnClass` to check this?

Reply:
sounds a good point.

Reply:
`scan` may throw `NotFoundException`, here we do not reference to `Conditional` because we treat it as somewhat Fatal error. There is no persistent class for database, the user may need to check the code.

Reply:
How other spring data project handle such scenario? Relying on Exception seems inelegant.

Reply:
Almost the same, in our `cosmosdb` it catch the exception, in `mongodb` it just throws the exception. Both are relying on Exceptions.

Reply:
I think using exception to manipulate app logic isn't the right practice. But since in this case you want to notify the user of failure, it makes sense to rethrow.

Reply:
agree, just trade off here if there is no better choice.